File: ports/typescript/src/blake3-keyed.ts
import { hash } from 'blake3';

/**
 * BLAKE3 keyed mode implementation
 * Uses BLAKE3 in keyed hashing mode with a hashed key
 */
function blake3KeyedMode(key: Uint8Array, data: Uint8Array, outLen: number): Uint8Array {
  // Hash the key first to get a 32-byte key
  const hashedKey = hash(key);
  
  // Use BLAKE3 keyed hash with the hashed key
  const hasher = hash(data, { key: hashedKey });
  
  // Return the specified output length
  return hasher.slice(0, outLen);
}

/**
 * BLAKE3 keyed mode with 256-bit (32 bytes) output
 */
export function blake3KeyedMode256(key: Uint8Array, data: Uint8Array): Uint8Array {
  return blake3KeyedMode(key, data, 32);
}

/**
 * BLAKE3 keyed mode with 512-bit (64 bytes) output
 */
export function blake3KeyedMode512(key: Uint8Array, data: Uint8Array): Uint8Array {
  return blake3KeyedMode(key, data, 64);
}

File: ports/rust/src/blake3_keyed.rs
use blake3::Hasher;

/// BLAKE3 keyed mode implementation
fn blake3_keyed_mode(key: &[u8], data: &[u8], out_len: usize) -> Vec<u8> {
    // Hash the key first to get a 32-byte key
    let hashed_key = blake3::hash(key);
    
    // Use BLAKE3 keyed hash with the hashed key
    let mut hasher = Hasher::new_keyed(hashed_key.as_bytes());
    hasher.update(data);
    
    // Get the output with specified length
    let mut output = vec![0u8; out_len];
    let mut reader = hasher.finalize_xof();
    reader.fill(&mut output);
    output
}

/// BLAKE3 keyed mode with 256-bit (32 bytes) output
pub fn blake3_keyed_mode256(key: &[u8], data: &[u8]) -> Vec<u8> {
    blake3_keyed_mode(key, data, 32)
}

/// BLAKE3 keyed mode with 512-bit (64 bytes) output
pub fn blake3_keyed_mode512(key: &[u8], data: &[u8]) -> Vec<u8> {
    blake3_keyed_mode(key, data, 64)
}
